src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Completion/MarkupTransitionCompletionItemProvider.cs

```diff
@@ -0,0 +1,82 @@
+using Microsoft.AspNetCore.Razor.Language.Syntax;
+using RazorSyntaxNode = Microsoft.AspNetCore.Razor.Language.Syntax.SyntaxNode;
+
+namespace Microsoft.CodeAnalysis.Razor.Completion
+{
+    internal class MarkupTransitionCompletionItemProvider : RazorCompletionItemProvider
+    {
+        private static readonly IReadOnlyCollection<string> ElementCommitCharacters = new HashSet<string>{">"};
+        private readonly HtmlFactsService _htmlFactsService;
+
+        public MarkupTransitionCompletionItemProvider(HtmlFactsService htmlFactsService)
+        {
+            if (htmlFactsService is null)
+            {
+                throw new ArgumentNullException(nameof(htmlFactsService));
+            }
+
+            _htmlFactsService = htmlFactsService;
+        }
+
+        public override IReadOnlyList<RazorCompletionItem> GetCompletionItems(RazorSyntaxTree syntaxTree, TagHelperDocumentContext tagHelperDocumentContext, SourceSpan location)
+        {
+            if (syntaxTree is null)
+            {
+                throw new ArgumentNullException(nameof(syntaxTree));
+            }
+
+            var change = new SourceChange(location, string.Empty);
+            var owner = syntaxTree.Root.LocateOwner(change);
+
+            if (owner == null)
+            {
+                Debug.Fail("Owner should never be null.");
+                return Array.Empty<RazorCompletionItem>();
+            }
+
+            if (!AtMarkupTransitionCompletionPoint(owner))
+            {
+                return Array.Empty<RazorCompletionItem>();
+            }
+
+            var parent = owner.Parent;
+            if (!_htmlFactsService.TryGetElementInfo(parent, out var containingTagNameToken, out var attributes) ||
+                !containingTagNameToken.Span.IntersectsWith(location.AbsoluteIndex))
+            {
+                return Array.Empty<RazorCompletionItem>();
+            }
+
+            var completions = new List<RazorCompletionItem>();
+            var completionDisplayText = SyntaxConstants.TextTagName;
+            var completionItem = new RazorCompletionItem(
+                completionDisplayText,
+                completionDisplayText,
+                RazorCompletionItemKind.MarkupTransition,
+                ElementCommitCharacters);
+            var completionDescription = new MarkupTransitionCompletionDescription(CSharpCodeParser.AddMarkupTransitionDescriptor);
+            completionItem.SetMarkupTransitionCompletionDescription(completionDescription);
+            completions.Add(completionItem);
+
+            return completions;
+        }
+
+        // Internal for testing
+        internal static bool AtMarkupTransitionCompletionPoint(RazorSyntaxNode owner)
+        {
+            // Only provide IntelliSense for C# code blocks, of the form:
+            // @{ }, @code{ }, @functions{ }, @if(true){ }
+            var closestBlockLanguage = owner.FirstAncestorLanguageForMarkupTransition();
```

Comment:
We don't want methods that operate on specific Syntax types directly on the abstract class. Instead you can use the `Ancestors()` method like the following,
```C#
var closestSignificantAncestor = owner.Ancestors().FirstOrDefault(node =>
{
    return (node is MarkupElementSyntax && markupNode.ChildNodes().Count != 1) || node is CSharpCodeBlockSyntax;
});
return closestSignificantAncestor is CSharpCodeBlockSyntax;
```

Reply:
Thanks, that's a great point. I've updated accordingly.